File: provision/go/internal/environment/memory_output.go
package environment

import "sync"

type MemoryOutput struct {
	mu     sync.Mutex
	buffer string
}

func NewMemoryOutput() *MemoryOutput {
	return &MemoryOutput{}
}

func (output *MemoryOutput) Write(content string) {
	output.mu.Lock()
	defer output.mu.Unlock()

	output.buffer += content
}

func (output *MemoryOutput) Clear() {
	output.mu.Lock()
	defer output.mu.Unlock()

	output.buffer = ""
}

func (output *MemoryOutput) Text() string {
	output.mu.Lock()
	defer output.mu.Unlock()

	return output.buffer
}

File: pkgs-ts/console/src/environment/memory-output.ts
import { parsePromptOutputContent } from '#console/environment/validators/output';
import type { PromptOutput } from '#console/types';

export type MemoryOutput = PromptOutput & {
	clear(): void;
	text(): string;
};

export const createMemoryOutput = (): MemoryOutput => {
	let buffer = '';

	return {
		clear(): void {
			buffer = '';
		},
		write(content: string): void {
			buffer += parsePromptOutputContent(content);
		},
		text(): string {
			return buffer;
		},
	};
};
